src/parse.c

```diff
@@ -1105,12 +1105,76 @@ handle_addresses(yaml_document_t* doc, yaml_node_t* node, const void* _, GError*
     return handle_generic_addresses(doc, node, TRUE, &(cur_netdef->ip4_addresses), &(cur_netdef->ip6_addresses), error);
 }
 
+const static char *GW_MULTIPLE_IP_ERROR = "gateway: there must be at most one of each IP type";
+
+struct gw_parse_result {
+	const char *gw4;
+	const char *gw6;
+};
+
+static gboolean
+handle_single_gateway(yaml_node_t* node, GError** error, struct gw_parse_result *out)
+{
+    const char *addr;
+
+    assert_type(node, YAML_SCALAR_NODE);
+
+    addr = scalar(node);
+    if (is_ip4_address(addr)) {
+	    if (out->gw4)
+		    return yaml_error(node, error, GW_MULTIPLE_IP_ERROR);
+	    out->gw4 = addr;
+    } else if (is_ip6_address(addr)) {
+	    if (out->gw6)
+		    return yaml_error(node, error, GW_MULTIPLE_IP_ERROR);
+	    out->gw6 = addr;
+    } else
+	    return yaml_error(node, error, "invalid IP address '%s'", addr);
+    return TRUE;
+}
+
+static gboolean
+handle_gateway(yaml_document_t* doc, yaml_node_t* node, const void* _, GError** error)
+{
+    struct gw_parse_result res = {NULL, NULL};
+
+    if (node->type == YAML_SCALAR_NODE) {
+        if (!handle_single_gateway(node, error, &res))
+            return FALSE;
+    } else {
+        yaml_node_item_t *iter;
+        assert_type(node, YAML_SEQUENCE_NODE);
+
+        for (iter = node->data.sequence.items.start; iter < node->data.sequence.items.top; iter++) {
+            if (!handle_single_gateway(yaml_document_get_node(doc, *iter), error, &res))
+                return FALSE;
+        }
+    }
+
+    if (res.gw4) {
+        if (cur_netdef->gateway4)
+            g_assert_cmpstr(cur_netdef->gateway4, ==, res.gw4);
+        else
+            cur_netdef->gateway4 = g_strdup(res.gw4);
+    }
+
+    if (res.gw6) {
+        if (cur_netdef->gateway6)
+            g_assert_cmpstr(cur_netdef->gateway6, ==, res.gw6);
+        else
+            cur_netdef->gateway6 = g_strdup(res.gw6);
+    }
+
```

Comment:
Just for clarity:The `res.gw4/6` values are still the same from inside the YAML `node`, right? That's why we don't need to `g_free()` them down here?